From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stefano Stabellini <sstabellini@kernel.org>
Date: Tue, 26 Feb 2019 15:00:28 -0800
Subject: [PATCH] libxl/xl: add cacheability option to iomem

Parse a new cacheability option for the iomem parameter, it can be
"devmem" for device memory mappings, which is the default, or "memory"
for normal memory mappings.

Store the parameter in a new field in libxl_iomem_range.

Pass the cacheability option to xc_domain_memory_mapping.

Signed-off-by: Stefano Stabellini <stefanos@xilinx.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
(cherry picked from commit d4823d5ef97e69e45e97c1a8fd768427df43eec0)
---
 docs/man/xl.cfg.5.pod.in         |  4 ++-
 tools/libs/light/libxl_create.c  | 12 +++++++-
 tools/libs/light/libxl_types.idl |  7 +++++
 tools/xl/xl_parse.c              | 49 ++++++++++++++++++++++++++++----
 4 files changed, 64 insertions(+), 8 deletions(-)

diff --git a/docs/man/xl.cfg.5.pod.in b/docs/man/xl.cfg.5.pod.in
index 85dfdd39e9..566d9e162f 100644
--- a/docs/man/xl.cfg.5.pod.in
+++ b/docs/man/xl.cfg.5.pod.in
@@ -1342,7 +1342,7 @@ is given in hexadecimal format and may either be a range, e.g. C<2f8-2ff>
 It is recommended to only use this option for trusted VMs under
 administrator's control.
 
-=item B<iomem=[ "IOMEM_START,NUM_PAGES[@GFN]", "IOMEM_START,NUM_PAGES[@GFN]", ...]>
+=item B<iomem=[ "IOMEM_START,NUM_PAGES[@GFN][,CACHEABILITY]", "IOMEM_START,NUM_PAGES[@GFN][,CACHEABILITY]", ...]>
 
 Allow auto-translated domains to access specific hardware I/O memory pages.
 
@@ -1353,6 +1353,8 @@ B<GFN> is not specified, the mapping will be performed using B<IOMEM_START>
 as a start in the guest's address space, therefore performing a 1:1 mapping
 by default.
 All of these values must be given in hexadecimal format.
+B<CACHEABILITY> can be "devmem" for device memory, the default if not
+specified, or it can be "memory" for normal memory.
 
 Note that the IOMMU won't be updated with the mappings specified with this
 option. This option therefore should not be used to pass through any
diff --git a/tools/libs/light/libxl_create.c b/tools/libs/light/libxl_create.c
index e73c35edde..b71169e506 100644
--- a/tools/libs/light/libxl_create.c
+++ b/tools/libs/light/libxl_create.c
@@ -457,6 +457,14 @@ static void init_console_info(libxl__gc *gc,
        Only 'channels' when mapped to consoles have a string name. */
 }
 
+static uint32_t libxl__cacheability_to_xc(libxl_cacheability c)
+{
+    if (c == LIBXL_CACHEABILITY_MEMORY)
+        return CACHEABILITY_MEMORY;
+    /* default to devmem */
+    return CACHEABILITY_DEVMEM;
+}
+
 int libxl__domain_build(libxl__gc *gc,
                         libxl_domain_config *d_config,
                         uint32_t domid,
@@ -1736,7 +1744,9 @@ static void domcreate_launch_dm(libxl__egc *egc, libxl__multidev *multidev,
         }
         ret = xc_domain_memory_mapping_cache(CTX->xch, domid,
                                        io->gfn, io->start,
-                                       io->number, 1, CACHEABILITY_DEVMEM);
+                                       io->number, 1,
+                                       libxl__cacheability_to_xc(
+                                           io->cache_policy));
         if (ret < 0) {
             LOGED(ERROR, domid,
                   "failed to map to domain iomem range %"PRIx64"-%"PRIx64
diff --git a/tools/libs/light/libxl_types.idl b/tools/libs/light/libxl_types.idl
index 7b75d862e4..75e3b68b03 100644
--- a/tools/libs/light/libxl_types.idl
+++ b/tools/libs/light/libxl_types.idl
@@ -314,6 +314,11 @@ libxl_ioport_range = Struct("ioport_range", [
     ("number", uint32),
     ])
 
+libxl_cacheability = Enumeration("cacheability", [
+    (0, "devmem"),
+    (1, "memory"),
+    ], init_val = "LIBXL_CACHEABILITY_DEVMEM")
+
 libxl_iomem_range = Struct("iomem_range", [
     # start host frame number to be mapped to the guest
     ("start", uint64),
@@ -321,6 +326,8 @@ libxl_iomem_range = Struct("iomem_range", [
     ("number", uint64),
     # guest frame number used as a start for the mapping
     ("gfn", uint64, {'init_val': "LIBXL_INVALID_GFN"}),
+    # cacheability of the memory region
+    ("cache_policy", libxl_cacheability),
     ])
 
 libxl_vga_interface_info = Struct("vga_interface_info", [
diff --git a/tools/xl/xl_parse.c b/tools/xl/xl_parse.c
index d4fe4feb5a..524ec715e2 100644
--- a/tools/xl/xl_parse.c
+++ b/tools/xl/xl_parse.c
@@ -2390,6 +2390,7 @@ void parse_config_data(const char *config_source,
         }
         for (i = 0; i < num_iomem; i++) {
             int used;
+            const char *iter;
 
             buf = xlu_cfg_get_listitem (iomem, i);
             if (!buf) {
@@ -2398,20 +2399,56 @@ void parse_config_data(const char *config_source,
                 exit(1);
             }
             libxl_iomem_range_init(&b_info->iomem[i]);
-            ret = sscanf(buf, "%" SCNx64",%" SCNx64"%n@%" SCNx64"%n",
+            /* Parse mandatory part */
+            ret = sscanf(buf, "%" SCNx64",%" SCNx64"%n",
                          &b_info->iomem[i].start,
-                         &b_info->iomem[i].number, &used,
-                         &b_info->iomem[i].gfn, &used);
-            if (ret < 2 || buf[used] != '\0') {
+                         &b_info->iomem[i].number, &used);
+            if (ret < 2) {
                 fprintf(stderr,
                         "xl: Invalid argument parsing iomem: %s\n", buf);
                 exit(1);
             }
+            if (buf[used] == '\0')
+                continue;
+
+            /* Parse optional "@gfn" or ",cacheability" */
+            iter = buf + used;
+            if (iter[0] == '@') {
+                char *ep;
+
+                iter++;
+                b_info->iomem[i].gfn = strtoull(iter, &ep, 16);
+                if (ep == iter) {
+                    fprintf(stderr,
+                            "xl: Invalid @gfn argument in iomem: %s\n", buf);
+                    exit(1);
+                }
+
+                if (*ep == '\0')
+                    continue;
+
+                iter = ep;
+            }
+
+            if (iter[0] == ',') {
+                iter++;
+
+                if (!strncmp(iter, "memory", strlen(iter)))
+                    b_info->iomem[i].cache_policy = LIBXL_CACHEABILITY_MEMORY;
+                else if (!strncmp(iter, "devmem", strlen(iter)))
+                    b_info->iomem[i].cache_policy = LIBXL_CACHEABILITY_DEVMEM;
+                else {
+                    fprintf(stderr,
+                            "xl: Invalid iomem cache parameter: %s\n", buf);
+                    exit(1);
+                }
+            } else {
+                fprintf(stderr, "xl: Invalid iomem format: %s\n", buf);
+                exit(1);
+            }
         }
     }
 
-
-
     if (!xlu_cfg_get_list (config, "disk", &vbds, 0, 0)) {
         d_config->num_disks = 0;
         d_config->disks = NULL;
-- 
2.43.0

